### PR TITLE
Create property ZAI seam for PHP 8

### PR DESCRIPTION
**Pull request by bwoebi**
### Description

A simple function allowing direct access into userland (and internal classes without very special property storage) properties, bypassing __get() and similar.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.
- [x] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: zend_abstract_interface/properties/tests/php7-8/properties.cc [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        REQUIRE(zai_sapi_spinup()); \
+        ZAI_SAPI_TSRMLS_FETCH(); \
+        ZAI_SAPI_ABORT_ON_BAILOUT_OPEN() \
+        REQUIRE(zai_sapi_execute_script("./stubs/classes.php")); \
+        { code } \
+        ZAI_SAPI_ABORT_ON_BAILOUT_CLOSE() \
+        zai_sapi_spindown(); \
+    }
+
+static inline zval *zai_read_property_direct_literal(zend_class_entry *ce, zend_object *obj, const char *name, int name_len) {
```
Comment (morrisonlevi) [inline zend_abstract_interface/properties/tests/php7-8/properties.cc:20]: Nit: in C99 and newer there is no need for `inline` on a function in a source file, so our style is to omit it.

File: zend_abstract_interface/properties/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    set(PHP_PROPERTIES_VERSION_DIRECTORY PHP_VERSION_DIRECTORY)
+endif()
+
+add_library(zai_properties "${PHP_PROPERTIES_VERSION_DIRECTORY}/properties.c")
+
+target_include_directories(zai_properties PUBLIC
+                                          $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+                                          $<INSTALL_INTERFACE:include>)
+
+target_compile_features(zai_properties PUBLIC c_std_99)
```
Comment (morrisonlevi) [inline zend_abstract_interface/properties/CMakeLists.txt:13]: I was expecting this to be either `INTERFACE c_std_99 PRIVATE c_std_11` or `PUBLIC c_std_11` -- any specific reason for choosing C99?

File: zend_abstract_interface/properties/php7-8/properties.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#include "properties.h"
+
+#include <assert.h>
+#include <sandbox/sandbox.h>
```
Comment (SammyK) [inline zend_abstract_interface/properties/php7-8/properties.c:4]: I don't see any sandbox symbols being used. Was this left over from a refactor?

File: zend_abstract_interface/properties/CMakeLists.txt (hunk 2 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/CMakeLists.txt:13]: just copy paste from the other CMakeLists. Doesn't really matter as long as it compiles?

File: package.xml [hunk 4]
```diff
@@ -56,6 +56,9 @@
             <file name="zend_abstract_interface/sandbox/php7/sandbox.c" role="src" />
             <file name="zend_abstract_interface/sandbox/php8/sandbox.c" role="src" />
             <file name="zend_abstract_interface/sandbox/sandbox.h" role="src" />
+            <file name="zend_abstract_interface/properties/php7-8/properties.c" role="src" />
```
Comment (SammyK) [inline package.xml:59]: We'll also need to add `zend_abstract_interface/properties/php7-8/properties.c` to `ZAI_SOURCES` for PHP 8 in `config.m4`. And also the new build dirs:

```
PHP_ADD_BUILD_DIR([$ext_builddir/zend_abstract_interface/properties])
PHP_ADD_BUILD_DIR([$ext_builddir/zend_abstract_interface/properties/php7-8])
```

File: zend_abstract_interface/properties/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                                     VERSION ${PROJECT_VERSION})
+
+add_library(Zai::Properties ALIAS zai_properties)
+
+if (${BUILD_ZAI_TESTING})
+  add_subdirectory(tests)
+endif()
+
+# This copies the include files when `install` is ran
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/properties.h ${CMAKE_CURRENT_SOURCE_DIR}/properties_common.h
```
Comment (SammyK) [inline zend_abstract_interface/properties/CMakeLists.txt:28]: I'm guessing `properties_common.h` was left over from a refactor?

File: zend_abstract_interface/properties/tests/php7-8/properties.cc [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+#define TEST(name, code) TEST_CASE(name, "[zai properties]") { \
+        REQUIRE(zai_sapi_spinup()); \
+        ZAI_SAPI_TSRMLS_FETCH(); \
+        ZAI_SAPI_ABORT_ON_BAILOUT_OPEN() \
+        REQUIRE(zai_sapi_execute_script("./stubs/classes.php")); \
+        { code } \
+        ZAI_SAPI_ABORT_ON_BAILOUT_CLOSE() \
+        zai_sapi_spindown(); \
+    }
```
Comment (SammyK) [inline zend_abstract_interface/properties/tests/php7-8/properties.cc:18]: Cool macro! 😎

File: zend_abstract_interface/properties/tests/php7-8/properties.cc [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+static inline zval *zai_read_property_direct_literal(zend_class_entry *ce, zend_object *obj, const char *name, int name_len) {
+    zend_string *name_str = zend_string_init(name, name_len, 0);
+    zval *val = zai_read_property_direct(ce, obj, name_str);
+    zend_string_release(name_str);
+    return val;
+}
+#define zai_read_property_direct_literal(obj, ce, name) zai_read_property_direct_literal(obj, ce, ZEND_STRL(name))
```
Comment (SammyK) [inline zend_abstract_interface/properties/tests/php7-8/properties.cc:26]: I think this would be useful to make public in `properties.h` so that that caller doesn't have to worry about making `zend_strings`. I believe most (or all?) of the time we will use this API, it will be with a string literal. WDYT?

File: zend_abstract_interface/properties/tests/php7-8/properties.cc [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        REQUIRE(zai_sapi_execute_script("./stubs/classes.php")); \
+        { code } \
+        ZAI_SAPI_ABORT_ON_BAILOUT_CLOSE() \
+        zai_sapi_spindown(); \
+    }
+
+static inline zval *zai_read_property_direct_literal(zend_class_entry *ce, zend_object *obj, const char *name, int name_len) {
+    zend_string *name_str = zend_string_init(name, name_len, 0);
+    zval *val = zai_read_property_direct(ce, obj, name_str);
+    zend_string_release(name_str);
```
Comment (SammyK) [inline zend_abstract_interface/properties/tests/php7-8/properties.cc:23]: Do you think it's worth using the `ZSTR_ALLOCA_ALLOC` API here?

File: zend_abstract_interface/properties/tests/php7-8/properties.cc [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+TEST("access property with invalid name is error", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "\0with zero byte");
+    REQUIRE(val == &EG(error_zval));
+
+    zval_ptr_dtor(&child);
+})
```
Comment (SammyK) [inline zend_abstract_interface/properties/tests/php7-8/properties.cc:134]: In the spirit of being super defensive, could we add some tests where `ce`, `object`, and `name` args are `NULL`?

Also, do you think it's worth it to add some tests to access properties on instances of internal classes as well?

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 7 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:26]: Well, our _current_ uses do not have it as string literal, but always zend_string *. I don't mind adding it publicly though (at least for future use).

File: package.xml (hunk 4 above)
Comment (bwoebi) [reply package.xml:59]: I added that in a separate (yet to be committed branch), but will backport it here, makes sense…

File: zend_abstract_interface/properties/CMakeLists.txt (hunk 5 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/CMakeLists.txt:28]: Yes, surprised that cmake doesn't complain about it missing.

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 8 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:23]: yes, as we move it to public API, yes.

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 9 above)
Comment (morrisonlevi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:134]: I agree with ensuring the caller does this check in principle, but until it can be enforced by static analysis or something, I would prefer to be defensive. For you and I getting a sigsegv directly at the site is nice, but for customers and their customers it's not good at all, as crashes bring down the process.

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 7 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:26]: (no, I mean, our current uses are actual zend_string * like `ZSTR_MESSAGE` - but will do anyway)

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 9 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:134]: Actually, we probably can just log that instead? I just have no idea what the proper logging within ZAI would be for that? ZAI has no access to the standard ddtrace_log functions

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 7 above)
Comment (SammyK) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:26]: Ah yes, gotcha. Thanks for clarifying. Since the known-string API wasn't added until PHP 7.2, I think using a string literal in this case would be a good way to keep the API consistent across all PHP versions. It'll be great to not need so many preprocessor directives in the tracer code. :)

File: zend_abstract_interface/properties/tests/php7-8/properties.cc (hunk 9 above)
Comment (bwoebi) [reply zend_abstract_interface/properties/tests/php7-8/properties.cc:134]: Re internal class access: there is no difference in the zend_object * structure whether something is internal or not. It just might be that internal classes have some data attached and may define their own read_properties, but this function does not support that use case anyway.

File: zend_abstract_interface/properties/php7-8/properties.h [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#ifndef ZAI_PROPERTIES_H
+#define ZAI_PROPERTIES_H
+
+#include "Zend/zend_API.h"
+
+// Note that the function will completely bypass read_property. ce (scope) must be object ce or parent of object ce.
+zval *zai_read_property_direct(zend_class_entry *ce, zend_object *object, zend_string *name);
+
+zval *zai_read_property_direct_literal(zend_class_entry *ce, zend_object *obj, const char *name, int name_len);
+#define zai_read_property_direct_literal(obj, ce, name) zai_read_property_direct_literal(obj, ce, ZEND_STRL(name))
```
Comment (morrisonlevi) [inline zend_abstract_interface/properties/php7-8/properties.h:10]: What do you think about naming the function `zai_read_property_direct_cstr` and leaving the macro as `zai_read_property_direct_literal`? This way if I have a ptr + len pair already for some reason I can call the function directly without undef'ing the macro.

Comment (bwoebi) [reply zend_abstract_interface/properties/php7-8/properties.h:10]: I agree, I was already doubting that decision myself :-D

**Review by SammyK — APPROVED** [review]
Nice work @bwoebi! LGTM 👍
